File: concepts/binarytrees/flatten.py
import sys
sys.path.append("../..")
from debug import *

def flatten(node : TreeNode) -> TreeNode:
	######################
	# Flattens a binary tree in a singly linked list down its right children
	######################
	if not node:
		return

	if node.right:
		flatten(node.right)

	if node.left:
		flatten(node.left)
		curr = leftroot = node.left
		while curr.right:
			curr = curr.right
		curr.right = node.right
		node.left = None
		node.right = leftroot


t = createBinaryTree([1,2,5,3,4,None,6])
flatten(t)

print("")
curr = t
while curr:
	print(curr.val)
	curr = curr.right

File: concepts/binarytrees/flatten.cpp
//////////////////////////////////////////////////////
//  Author: Jacob Nardone
//  Date created:
//  File type: C++ source file
//  Description:
//////////////////////////////////////////////////////


#include <mystd.h>
using namespace std;



void flatten(TreeNode *root){
	if (root == nullptr){ return; }

	if (root->right){
		flatten(root->right);
	}

	if (root->left){
		flatten(root->left);
		TreeNode *right = root->right;
		root->right = root->left;
		root->left = nullptr;
		while (root->right) root = root->right;
		root->right = right;
	}
}



int main(){
	TreeNode *root = createBinaryTree("1,2,4,3,null");
	dlog(root);
	flatten(root);
	dlog(root);
	return 0;
}



